### ISSUE TITLE
Flask: allow instrumenting pre-request

### ISSUE THREAD

**Issue by eMerzh** [DataDog/dd-trace-py#653]
in #409 it seems we moved away from using signals to start the span.
but it seems that signals are called earlier and then should be used preferentially.
also it could be cool to allow a split between pre-request, request and - why not- templating.
it seems that newrelic apm is wrapping the preprocess_request , and that ES APM is using signals.

**Comment by brettlangdon** [comment]
Hey @eMerzh thank you for reaching out.

We have some changes coming to our Flask integration slotted for our next release. I will try to keep you in the loop once those changes are available so you can try them out to see if they meet your needs better.

**Comment by eMerzh** [comment]
awesome :) 
let me know, i'll try my best to help :)

**Comment by brettlangdon** [comment]
@eMerzh I just opened an initial pull request for my changes, #667 

I would love if you gave the sample traces in the description a look over to see if this implementation better fits your needs.

As well, the code appears stable right now, if you were interested in giving it a try:

```python
pip install git+https://github.com/datadog/dd-trace-py@brettlangdon/flask-dev
```

**Comment by eMerzh** [comment]
those traces looks really much completed, and before / after requests are well tracked, so seems pretty good! 👍 
now , i'll try to test that, but unsure how i can put this in prod

**Comment by brettlangdon** [comment]
Hey @eMerzh awesome, thanks!

The easiest thing to do is just use `git+https://github.com/datadog/dd-trace-py@brettlangdon/flask-dev` in place of `ddtrace[==<version>]` in your `requirements.txt` or however you install `ddtrace` in production.

Please let me know if there is anything else I can do to help you test this!

**Comment by brettlangdon** [comment]
@eMerzh the Flask rewrite has been released with version `0.16.0` of `ddtrace` today.
